{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "2751e6e8-f0c4-4835-872a-0ffff459062a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from gtfparse import read_gtf\n",
    "import functools as ft\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ced7f349-2720-4cbf-b4d3-72ea5e3cd88e",
   "metadata": {},
   "source": [
    "## Genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "58780289-78c9-4617-ab38-7b98e0a955c4",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Extracted GTF attributes: ['gene_id', 'gene_type', 'gene_name', 'level', 'tag', 'transcript_id', 'transcript_type', 'transcript_name', 'transcript_support_level', 'havana_transcript', 'exon_number', 'exon_id', 'hgnc_id', 'havana_gene', 'ont', 'protein_id', 'ccdsid', 'artif_dupl']\n"
     ]
    }
   ],
   "source": [
    "# Read GTF with gene annotation\n",
    "genes = read_gtf('/DATA/users/magnitov/tacl/genome/gencode.v44.annotation.gtf')\n",
    "genes = genes[genes['feature'] == 'gene']\n",
    "genes = genes[['seqname', 'start', 'end', 'strand', 'gene_id', 'gene_type', 'gene_name']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "156eee93-decb-40dc-8291-bbcf800b721d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_counts_table(genes, sample, replicate):\n",
    "    counts_fwd = pd.read_csv('/DATA/users/magnitov/tacl/bruseq/counts/counts_forward.TACL_BrUseq.txt', \n",
    "                         names = ['gene_id', \n",
    "                                  'T-MAU2_NT_rep1', 'T-MAU2_NT_rep2', 'T-MAU2_NT_rep3', \n",
    "                                  'T-MAU2_Dox2h_rep1', 'T-MAU2_Dox2h_rep2', 'T-MAU2_Dox2h_rep3',\n",
    "                                  'T-mCherry_NT_rep1', 'T-mCherry_NT_rep2', 'T-mCherry_NT_rep3',\n",
    "                                  'T-mCherry_Dox2h_rep1', 'T-mCherry_Dox2h_rep2', 'T-mCherry_Dox2h_rep3'],\n",
    "                         sep = '\\t', header = None)\n",
    "    counts_fwd = counts_fwd[counts_fwd['gene_id'].isin([x for x in counts_fwd['gene_id'] if x[0]!='_'])]\n",
    "    counts_fwd = counts_fwd[['gene_id', sample + '_' + replicate]]\n",
    "    \n",
    "    counts_rev = pd.read_csv('/DATA/users/magnitov/tacl/bruseq/counts/counts_reverse.TACL_BrUseq.txt', \n",
    "                         names = ['gene_id', \n",
    "                                  'T-MAU2_NT_rep1', 'T-MAU2_NT_rep2', 'T-MAU2_NT_rep3', \n",
    "                                  'T-MAU2_Dox2h_rep1', 'T-MAU2_Dox2h_rep2', 'T-MAU2_Dox2h_rep3',\n",
    "                                  'T-mCherry_NT_rep1', 'T-mCherry_NT_rep2', 'T-mCherry_NT_rep3',\n",
    "                                  'T-mCherry_Dox2h_rep1', 'T-mCherry_Dox2h_rep2', 'T-mCherry_Dox2h_rep3'],\n",
    "                         sep = '\\t', header = None)\n",
    "    counts_rev = counts_rev[counts_rev['gene_id'].isin([x for x in counts_rev['gene_id'] if x[0]!='_'])]\n",
    "    counts_rev = counts_rev[['gene_id', sample + '_' + replicate]]\n",
    "    \n",
    "    genes_merged = genes.merge(counts_fwd, on = 'gene_id', how = 'left')\n",
    "    genes_merged = genes_merged.merge(counts_rev, on = 'gene_id', how = 'left')\n",
    "    \n",
    "    counts_new = []\n",
    "    for gene in genes_merged.values:\n",
    "        if gene[3] == '+':\n",
    "            counts_new.append(gene[7])\n",
    "        else:\n",
    "            counts_new.append(gene[8])\n",
    "    genes_merged[f'{sample}_{replicate}'] = counts_new\n",
    "    \n",
    "    genes_merged = genes_merged.fillna(0)\n",
    "    \n",
    "    genes_merged = genes_merged[['seqname', 'start', 'end', 'strand', 'gene_id', 'gene_type', 'gene_name', f'{sample}_{replicate}']]\n",
    "    \n",
    "    return(genes_merged)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "4158e825-0815-4243-9b7f-f3956e018a7d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>T-MAU2_NT_rep1</th>\n",
       "      <th>T-MAU2_NT_rep2</th>\n",
       "      <th>T-MAU2_NT_rep3</th>\n",
       "      <th>T-MAU2_Dox2h_rep1</th>\n",
       "      <th>T-MAU2_Dox2h_rep2</th>\n",
       "      <th>T-MAU2_Dox2h_rep3</th>\n",
       "      <th>T-mCherry_NT_rep1</th>\n",
       "      <th>T-mCherry_NT_rep2</th>\n",
       "      <th>T-mCherry_NT_rep3</th>\n",
       "      <th>T-mCherry_Dox2h_rep1</th>\n",
       "      <th>T-mCherry_Dox2h_rep2</th>\n",
       "      <th>T-mCherry_Dox2h_rep3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000290825.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>DDX11L2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12010</td>\n",
       "      <td>13670</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000223972.6</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14404</td>\n",
       "      <td>29570</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>47</td>\n",
       "      <td>22</td>\n",
       "      <td>31</td>\n",
       "      <td>22</td>\n",
       "      <td>28</td>\n",
       "      <td>31</td>\n",
       "      <td>10</td>\n",
       "      <td>30</td>\n",
       "      <td>37</td>\n",
       "      <td>26</td>\n",
       "      <td>12</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>17369</td>\n",
       "      <td>17436</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>29554</td>\n",
       "      <td>31109</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqname  start    end strand            gene_id  \\\n",
       "0    chr1  11869  14409      +  ENSG00000290825.1   \n",
       "1    chr1  12010  13670      +  ENSG00000223972.6   \n",
       "2    chr1  14404  29570      -  ENSG00000227232.5   \n",
       "3    chr1  17369  17436      -  ENSG00000278267.1   \n",
       "4    chr1  29554  31109      +  ENSG00000243485.5   \n",
       "\n",
       "                            gene_type    gene_name  T-MAU2_NT_rep1  \\\n",
       "0                              lncRNA      DDX11L2               0   \n",
       "1  transcribed_unprocessed_pseudogene      DDX11L1               0   \n",
       "2              unprocessed_pseudogene       WASH7P              47   \n",
       "3                               miRNA    MIR6859-1               7   \n",
       "4                              lncRNA  MIR1302-2HG               0   \n",
       "\n",
       "   T-MAU2_NT_rep2  T-MAU2_NT_rep3  T-MAU2_Dox2h_rep1  T-MAU2_Dox2h_rep2  \\\n",
       "0               0               1                  0                  0   \n",
       "1               0               0                  0                  0   \n",
       "2              22              31                 22                 28   \n",
       "3               1               2                  3                  4   \n",
       "4               0               1                  0                  0   \n",
       "\n",
       "   T-MAU2_Dox2h_rep3  T-mCherry_NT_rep1  T-mCherry_NT_rep2  T-mCherry_NT_rep3  \\\n",
       "0                  0                  0                  0                  0   \n",
       "1                  0                  0                  0                  0   \n",
       "2                 31                 10                 30                 37   \n",
       "3                  4                  0                  4                  0   \n",
       "4                  0                  0                  0                  0   \n",
       "\n",
       "   T-mCherry_Dox2h_rep1  T-mCherry_Dox2h_rep2  T-mCherry_Dox2h_rep3  \n",
       "0                     0                     0                     0  \n",
       "1                     0                     0                     0  \n",
       "2                    26                    12                    10  \n",
       "3                     3                     1                     1  \n",
       "4                     0                     0                     0  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts_T_MAU2_rep1 = get_counts_table(genes, 'T-MAU2_NT', 'rep1')\n",
    "counts_T_MAU2_rep2 = get_counts_table(genes, 'T-MAU2_NT', 'rep2')\n",
    "counts_T_MAU2_rep3 = get_counts_table(genes, 'T-MAU2_NT', 'rep3')\n",
    "counts_T_mCherry_rep1 = get_counts_table(genes, 'T-mCherry_NT', 'rep1')\n",
    "counts_T_mCherry_rep2 = get_counts_table(genes, 'T-mCherry_NT', 'rep2')\n",
    "counts_T_mCherry_rep3 = get_counts_table(genes, 'T-mCherry_NT', 'rep3')\n",
    "counts_T_MAU2_Dox_rep1 = get_counts_table(genes, 'T-MAU2_Dox2h', 'rep1')\n",
    "counts_T_MAU2_Dox_rep2 = get_counts_table(genes, 'T-MAU2_Dox2h', 'rep2')\n",
    "counts_T_MAU2_Dox_rep3 = get_counts_table(genes, 'T-MAU2_Dox2h', 'rep3')\n",
    "counts_T_mCherry_Dox_rep1 = get_counts_table(genes, 'T-mCherry_Dox2h', 'rep1')\n",
    "counts_T_mCherry_Dox_rep2 = get_counts_table(genes, 'T-mCherry_Dox2h', 'rep2')\n",
    "counts_T_mCherry_Dox_rep3 = get_counts_table(genes, 'T-mCherry_Dox2h', 'rep3')\n",
    "\n",
    "count_tables = [counts_T_MAU2_rep1, counts_T_MAU2_rep2, counts_T_MAU2_rep3, \n",
    "                counts_T_MAU2_Dox_rep1, counts_T_MAU2_Dox_rep2, counts_T_MAU2_Dox_rep3,\n",
    "                counts_T_mCherry_rep1, counts_T_mCherry_rep2, counts_T_mCherry_rep3,\n",
    "                counts_T_mCherry_Dox_rep1, counts_T_mCherry_Dox_rep2, counts_T_mCherry_Dox_rep3]\n",
    "counts_tacl = ft.reduce(lambda left, right: pd.merge(left, right, on = ['seqname', 'start', 'end', 'strand', 'gene_id', 'gene_type', 'gene_name']), count_tables)\n",
    "for column in counts_tacl.columns[7:]:\n",
    "    counts_tacl[column] = [int(x) for x in counts_tacl[column]]\n",
    "counts_tacl.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "a8fa6d1f-865b-4243-9126-c8aa46da7b10",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>T-MAU2_NT_rep1</th>\n",
       "      <th>T-MAU2_NT_rep2</th>\n",
       "      <th>T-MAU2_NT_rep3</th>\n",
       "      <th>T-MAU2_Dox2h_rep1</th>\n",
       "      <th>T-MAU2_Dox2h_rep2</th>\n",
       "      <th>T-MAU2_Dox2h_rep3</th>\n",
       "      <th>T-mCherry_NT_rep1</th>\n",
       "      <th>T-mCherry_NT_rep2</th>\n",
       "      <th>T-mCherry_NT_rep3</th>\n",
       "      <th>T-mCherry_Dox2h_rep1</th>\n",
       "      <th>T-mCherry_Dox2h_rep2</th>\n",
       "      <th>T-mCherry_Dox2h_rep3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4016</th>\n",
       "      <td>chr1</td>\n",
       "      <td>169662007</td>\n",
       "      <td>169854080</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000000460.17</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>C1orf112</td>\n",
       "      <td>2061</td>\n",
       "      <td>1331</td>\n",
       "      <td>2117</td>\n",
       "      <td>1098</td>\n",
       "      <td>1504</td>\n",
       "      <td>2040</td>\n",
       "      <td>705</td>\n",
       "      <td>2013</td>\n",
       "      <td>2158</td>\n",
       "      <td>291</td>\n",
       "      <td>434</td>\n",
       "      <td>1094</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     seqname      start        end strand             gene_id       gene_type  \\\n",
       "4016    chr1  169662007  169854080      +  ENSG00000000460.17  protein_coding   \n",
       "\n",
       "     gene_name  T-MAU2_NT_rep1  T-MAU2_NT_rep2  T-MAU2_NT_rep3  \\\n",
       "4016  C1orf112            2061            1331            2117   \n",
       "\n",
       "      T-MAU2_Dox2h_rep1  T-MAU2_Dox2h_rep2  T-MAU2_Dox2h_rep3  \\\n",
       "4016               1098               1504               2040   \n",
       "\n",
       "      T-mCherry_NT_rep1  T-mCherry_NT_rep2  T-mCherry_NT_rep3  \\\n",
       "4016                705               2013               2158   \n",
       "\n",
       "      T-mCherry_Dox2h_rep1  T-mCherry_Dox2h_rep2  T-mCherry_Dox2h_rep3  \n",
       "4016                   291                   434                  1094  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts_tacl[counts_tacl['gene_id'] == 'ENSG00000000460.17']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "434f2274-2325-47ca-8844-3649b0fb7e1b",
   "metadata": {},
   "outputs": [],
   "source": [
    "counts_tacl.drop(['seqname', 'start', 'end', 'strand', 'gene_type', 'gene_name'], axis = 1).to_csv('/DATA/users/magnitov/tacl/bruseq/counts/counts_deseq2.TACL.txt', sep = '\\t', header = 0, index = 0)\n",
    "counts_tacl.to_csv('/DATA/users/magnitov/tacl/bruseq/counts/counts_analysis.TACL.txt', sep = '\\t', header = 1, index = 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6698fabe-7e4c-439f-a93f-6611bc516412",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:anaconda3]",
   "language": "python",
   "name": "conda-env-anaconda3-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
